crypto/crypto.c

```diff
@@ -0,0 +1,425 @@
+{
+        mbedtls_pkcs7_free((struct mbedtls_pkcs7 *)pkcs7);
+        free(pkcs7);
+}
+
+void* crypto_get_signing_cert(void *pkcs7, int cert_num)
+{
+    mbedtls_x509_crt *pkcs7_cert = NULL;
+
+    pkcs7_cert = &((struct mbedtls_pkcs7 *)pkcs7)->signed_data.certs;
+    for (int i = 0; i < cert_num && pkcs7_cert != NULL; i++)
+        pkcs7_cert = pkcs7_cert->next;
+
+    return (void *)pkcs7_cert;
+
+}
+
+int crypto_get_x509_der_len(void *x509) 
+{
+    return ((mbedtls_x509_crt *)x509)->raw.len; 
+}
+
+int crypto_get_tbs_x509_der_len(void *x509) 
+{
+    return ((mbedtls_x509_crt *)x509)->tbs.len;
+}
+
+int crypto_get_x509_version(void *x509) 
+{
+    return ((mbedtls_x509_crt *)x509)->version;
+}
+
+int crypto_x509_is_RSA(void *x509)
+{
+    int rc;
+    rc = ((mbedtls_x509_crt *)x509)->pk.pk_info->type;
+    if (rc != MBEDTLS_PK_RSA)
+        //zero is also a pk type (MBEDTLS_PK_NONE) so return generic failure if zero so it doesnt look like a success
+       return  (rc == 0 ? CERT_FAIL : rc);
+    else 
+        return SUCCESS;
+}
+
+int crypto_get_x509_sig_len(void *x509)
+{
+    return ((mbedtls_x509_crt *)x509)->sig.len;
+}
+
+int crypto_x509_md_is_sha256(void *x509)
+{
+    if (((mbedtls_x509_crt *)x509)->sig_md == MBEDTLS_MD_SHA256) 
+        return SUCCESS;
+    else
+        return CERT_FAIL;
+}
+
+int crypto_x509_oid_is_pkcs1_sha256(void *x509)
+{
+    if (strncmp((const char *)((mbedtls_x509_crt *)x509)->sig_oid.p, MBEDTLS_OID_PKCS1_SHA256, ((mbedtls_x509_crt *)x509)->sig_oid.len) )
+        return CERT_FAIL;
+    return SUCCESS;
+}
+
+int crypto_x509_get_pk_bit_len(void *x509)
+{
+    return mbedtls_pk_get_bitlen( &((mbedtls_x509_crt *)x509)->pk);
+}
+
+void crypto_x509_get_short_info(void *x509, char *short_desc, size_t max_len)
+{
+
+    mbedtls_x509_sig_alg_gets(short_desc, max_len, &((mbedtls_x509_crt *)x509)->sig_oid,
+                       ((mbedtls_x509_crt *)x509)->sig_pk, ((mbedtls_x509_crt *)x509)->sig_md, 
+                       ((mbedtls_x509_crt *)x509)->sig_opts );
+}
+
+int crypto_x509_get_long_desc(char *x509_info, size_t max_len, char *delim, void *x509)
+{
+    return mbedtls_x509_crt_info(x509_info, max_len, delim, ((mbedtls_x509_crt *)x509)); 
+}
+
+void *crypto_x509_parse_der(const unsigned char *data, size_t data_len)
+{
+    int rc;
+    mbedtls_x509_crt *x509 = NULL;
+    x509 = malloc(sizeof(*x509));
+    if (!x509){
+      prlog(PR_ERR, "ERROR: failed to allocate memory\n");
+      return NULL;
+    }
+    mbedtls_x509_crt_init(x509);
+    rc = mbedtls_x509_crt_parse(x509, data, data_len); 
+
+    if (rc) {
+        crypto_x509_free(x509);
+        return NULL;
+    }
+    else
+        return (void *)x509;
+}
+
+void crypto_x509_free(void *x509)
+{
+    mbedtls_x509_crt_free((mbedtls_x509_crt *)x509);
+    free(x509);
+}
+
+int crypto_convert_pem_to_der(const unsigned char *input, size_t ilen, unsigned char **output, size_t *olen)
+{
+    return convert_pem_to_der(input, ilen, output, olen);
+}
+#endif
```

Comment:
Abstractions look generally sane. One thing I would suggest: consider breaking the two implementations into two different files like `crypto-openssl.c` and `crypto-mbedtls.c`, and only building the file for the crypto library in use. Long `#ifdef`s make it a bit difficult to know which "mode" you are looking at (although, I suppose you could use context clues of function names to figure it out).